proposals/client-interceptors.md

```diff
@@ -0,0 +1,312 @@
+          var retry = function(request, metadata) {
+            retries++;
+            var newCall = nextCall(options);
+            newCall.start(metadata, {
+              onData: function(response) {
+                savedResponse = response;
+              },
+              onStatus: function(status) {
+                if (status.code !== StatusCode.OK) {
+                  if (retries <= maxRetries) {
+                    retry(request, metadata);
+                  } else {
+                    savedNext(savedResponse);
+                    next(status);
+                  }
+                } else {
+                  savedNext(savedResponse);
+                  next({code: StatusCode.OK})
+                }
+              },
+            })
+          }
+          if (status.code !== StatusCode.OK) {
+            retry(savedRequest, savedMetadata);
+          } else {
+            savedNext(savedResponse);
+            next(status);
+          }
+        }
+      };
+      next(metadata, listener);
+    },
+    send: function(request, next) {
+      savedRequest = request;
+      next(request);
+    },
+  };
+  return new InterceptingCall(nextCall(options), requester);
+}
+```
+
+### Usage
+
+Interceptors can be configured during client construction, or on individual invocations of gRPC calls.
+
+#### At client construction
+
+An `InterceptorProvider` type will be provided for computing the association of interceptors with gRPC calls dynamically.
+
+```javascript
+/**
+* @param {Function} getInterceptorForMethod A filter method which accepts a MethodDescriptor and returns
+* `undefined` (when no interceptor should be associated) or a single interceptor object.
+* @constructor
+*/
+InterceptorProvider(getInterceptorForMethod)
+```
+An array of `interceptors` can be passed in the `options` parameter when constructing a client:
+
+```javascript
+var client = new MyServiceClient(hostname, credentials, { interceptors: [ myInterceptor ] });
+```
+
+Alternatively, an array of `interceptorProviders` can be passed in the `options` parameter when constructing a client:
+
+```javascript
+var interceptorProviders = [
+    new InterceptorProvider(function(methodDescriptor) {
+        if (methodDescriptor.methodType === MethodType.UNARY) {
+            return unaryInterceptor;
+        }
+    }),
+    new InterceptorProvider(function(methodDescriptor) {
+        if (methodDescriptor.methodType === MethodType.SERVER_STREAMING) {
+            return streamingInterceptor;
+        }
+    }),
+];
+var constructorOptions = {
+    interceptorProviders: interceptorProviders,
+};
+var client = new MyServiceClient(hostname, credentials, constructorOptions);
+```
+The order of the InterceptorProviders will determine the order of the resulting interceptor stack.
+
+#### At call invocation
+
+If an array of interceptors is provided at call invocation via `options.interceptors`, the call will ignore all interceptors provided via the client constructor.
```

Comment:
Why ignore vS add to end of list? Do you have some concerns about interceptor ordering?

Interceptors like authorization, logging, metrics collection are likely to be defined and configured centrally. It would be unfortunate if a call site that just wants to add a custom interceptor needs to redefine this entire list again.

Reply:
I would expect that most interceptor authors to have there interceptor chains easily available to add at call invocation. 
Yes, the main reason is the ordering, or possibly you want a different `authorization` mechanism for that call etc.
I still think that this API has to override those at the client level, however maybe there is a second API that allows you to `add` interceptors to the chain? This would be useful for tools to inject interceptors; for example the grpc-web chrome DevTools.